pkg/controllers/util/listwatcher/listwatcher.go

```diff
@@ -0,0 +1,239 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/clock"
+)
+
+// ListWatcher lists and watches a given resource, emitting events to registered receivers. This allows receivers to
+// track changes to a resource starting from a consistent snapshot.
+//
+// Before any list request, ListWatcher must emit a StartSync event to notify receivers. This allows receivers to start
+// performing a resync if required. After all the objects from a successful list request is emitted, ListWatcher must
+// also emit a EndSync event to notify receivers.
+//
+// The design of ListWatcher and its implementations is loosely based on client-go's reflector.
+type ListWatcher interface {
+	Start(ctx context.Context)
+	AddReceiver(chan<- Event)
+}
+
+type Event struct {
+	Type   EventType
+	Object runtime.Object
+}
+
+type EventType string
+
+const (
+	// StartSync is emitted when a ListWatcher starts a new list. Receivers should use this as a signal to start a
+	// resync should they require a consistent view of the resource.
+	StartSync EventType = "StartSync"
+	// EndSync is emitted when a ListWatcher successfully completes a list. Receivers can use all the Add events
+	// received since the last StartSync to construct a consistent snapshot of the resource.
+	EndSync EventType = "EndSync"
+
+	// Add is emitted for each object in a list request, and any subsequent Added watch events.
+	Add EventType = "Add"
+	// Update is emitted for any Modified watch events.
+	Update EventType = "Update"
+	// Delete is emitted for any Deleted watch events.
+	Delete EventType = "Delete"
+)
+
+type pagedListWatcher struct {
+	lw        cache.ListerWatcher
+	receivers []chan<- Event
+	pageSize  int64
+
+	mu      sync.Mutex
+	mux     chan Event
+	backoff wait.BackoffManager
+}
+
+// NewPagedListWatcher returns a ListWatcher implementation that uses paging when listing from the apiserver. Each page
+// is processed immediately to allow the allocated memory to be potentially garbage collected before the next request.
+func NewPagedListWatcher(lw cache.ListerWatcher, pageSize int64) ListWatcher {
+	return &pagedListWatcher{
+		lw:        lw,
+		receivers: []chan<- Event{},
+		pageSize:  pageSize,
+		mu:        sync.Mutex{},
+		mux:       make(chan Event, 100),
+		backoff:   wait.NewExponentialBackoffManager(800*time.Millisecond, 30*time.Second, 2*time.Minute, 2.0, 1.0, &clock.RealClock{}),
+	}
+}
+
+func (p *pagedListWatcher) AddReceiver(receiver chan<- Event) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.receivers = append(p.receivers, receiver)
+}
+
+func (p *pagedListWatcher) Start(ctx context.Context) {
+	logger := klog.FromContext(ctx).WithValues("component", "paged-list-watcher")
+	listCh := make(chan string, 1)
+
+	go func() {
+		for {
+			// list resource
+			select {
+			case <-ctx.Done():
+				return
+			default:
```

Comment:
nit: we can move the main logic outside and select and eliminate some indentation